File: 2022/js/Day-04/Day-04.js
const fs = require("fs");

inputData = fs.readFileSync("input.txt", "utf8");
inputData = inputData
  .split(/\r?\n/)
  .map((line) =>
    line
      .split(",")
      .map((pair) => pair.split("-").map((number) => Number(number)))
  );

const Solution01 = (inputData) => {
  let counter = 0;
  for (pair of inputData) {
    if (
      (pair[0][0] >= pair[1][0] && pair[0][1] <= pair[1][1]) ||
      (pair[1][0] >= pair[0][0] && pair[1][1] <= pair[0][1])
    ) {
      counter++;
    }
  }
  return counter;
};

const Solution02 = (inputData) => {
  let counter = 0;
  for (pair of inputData) {
    if (
      (pair[0][0] >= pair[1][0] && pair[0][0] <= pair[1][1]) ||
      (pair[0][1] >= pair[1][0] && pair[0][1] <= pair[1][1]) ||
      (pair[1][0] >= pair[0][0] && pair[1][0] <= pair[0][1]) ||
      (pair[1][1] >= pair[0][0] && pair[1][1] <= pair[0][1])
    ) {
      counter++;
    }
  }
  return counter;
};

console.log(`Solution 01: ${Solution01(inputData)}`);
console.log(`Solution 02: ${Solution02(inputData)}`);

File: 2023/python/Day-04/Day_04.py
with open("input.txt", "r") as f:
    input_data = [
        [set([int(v) for v in numbers.split()]) for numbers in numbers]
        for numbers in [
            line.strip().split(": ")[1].split(" | ") for line in f.readlines()
        ]
    ]


def clamp_below_one(n):
    if n < 1:
        return 0
    return n


def solution01(cards):
    return sum(
        [
            clamp_below_one(2 ** (len(winning_numbers & my_numbers) - 1))
            for winning_numbers, my_numbers in cards
        ]
    )


def solution02(cards):
    cards_matching_numbers_n = [
        len(winning_numbers & my_numbers) for winning_numbers, my_numbers in cards
    ]

    cards_to_process = [i for i, _ in enumerate(cards)]

    processed_cards_counter = 0

    while cards_to_process:
        card = cards_to_process.pop()

        cards_to_process += [
            i for i in range(card + 1, card + 1 + cards_matching_numbers_n[card])
        ]

        processed_cards_counter += 1

    return processed_cards_counter


print("—")
print(f"Solution 01: {solution01(input_data)}")
print(f"Solution 02: {solution02(input_data)}")
print("—")
